docs/csharp/tutorials/pattern-matching.md

```diff
@@ -0,0 +1,325 @@
+
+You can test this code using the following code in `Program.cs`:
+
+```csharp
+using System;
+using CommercialRegistration;
+using ConsumerVehicleRegistration;
+using LiveryRegistration;
+
+namespace toll_calculator
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            var tollCalc = new TollCalculator();
+
+            var car = new Car();
+            var taxi = new Taxi();
+            var bus = new Bus();
+            var truck = new DeliveryTruck();
+
+            Console.WriteLine($"The toll for a car is {tollCalc.CalculateToll(car)}");
+            Console.WriteLine($"The toll for a taxi is {tollCalc.CalculateToll(taxi)}");
+            Console.WriteLine($"The toll for a bus is {tollCalc.CalculateToll(bus)}");
+            Console.WriteLine($"The toll for a truck is {tollCalc.CalculateToll(truck)}");
+
+            try
+            {
+                tollCalc.CalculateToll("this will fail");
+            }
+            catch (ArgumentException e)
+            {
+                Console.WriteLine("Caught an argument exception when using the wrong type", DayOfWeek.Friday);
+            }
+            try
+            {
+                tollCalc.CalculateToll(null);
+            }
+            catch (ArgumentNullException e)
+            {
+                Console.WriteLine("Caught an argument exception when using null");
+            }
+        }
+    }
+}
+```
+
+That code is included in the starter project, but is commented out. Remove the comments, and you can test what you've written.
+
+You're starting to see how patterns can help you create algorithms where the code and the data are separate. The `switch` expression tests the type and produces different values based on the results. That's only the beginning.
+
+## Add occupancy pricing
+
+The toll authority wants to encourage vehicles to travel at maximum capacity. They've decided to charge more when vehicles have fewer passengers, and encourage full vehicles by offering lower pricing:
+
+- Cars and taxis with no passengers pay an extra $0.50.
+- Cars and taxis with two passengers get a 0.50 discount.
+- Cars and taxis with three or more passengers get a $1.00 discount.
+- Buses that are less than 50% full pay an extra $2.00.
+- Buses that are more than 90% full get a $1.00 discount.
+
+These rules can be implemented using the **property pattern** in the same switch expression. The property pattern examines properties of the object once the type has been determined.  The single case for a `Car` expands to four different cases:
+
+```csharp
+Car { Passengers: 0} => 2.00m + 0.50m,
```

Comment:
I would put the whole switch statement in these code blocks (at least defining switch stuff) since this is teaching people how to use it, it would help them to see the context of these examples at all times.

Reply:
I added context on each of the snippets, then included the full method at the end of this section.